core/opendaq/reader/src/signal_reader.cpp

```diff
@@ -252,32 +237,89 @@ std::unique_ptr<Comparable> SignalReader::readStartDomain()
     return domainReader->readStart(domainPacket.getData(), info.prevSampleIndex, domainInfo);
 }
 
-void SignalReader::readUntilNextDataPacket()
+bool SignalReader::isFirstPacketEvent()
 {
     if (info.dataPacket.assigned())
     {
-        return;
+        return false;
+    }
+
+    auto packet = connection.peek();
+    while (packet.assigned())
+    {
+        if (packet.getType() == PacketType::Data)
+        {
+            connection.dequeue();
+            info.dataPacket = packet;
+            info.prevSampleIndex = 0;
+            return false;
+        }
+        else if (packet.getType() == PacketType::Event)
+        {
+            return true;
+        }
+        connection.dequeue();
     }
+    return false;
+}
+
+EventPacketPtr SignalReader::readUntilNextDataPacket()
+{
+    if (!isFirstPacketEvent())
+        return nullptr;
+
+    DataDescriptorPtr dataDescriptor;
+    DataDescriptorPtr domainDescriptor;
 
     PacketPtr packet = connection.dequeue();
-    while (packet.assigned() && packet.getType() != PacketType::Data)
+    while (packet.assigned() && (packet.getType() != PacketType::Data))
     {
-        bool firstData{false};
-        ErrCode errCode = handlePacket(packet, firstData);
-        checkErrorInfo(errCode);
+        if (packet.getType() == PacketType::Event)
+        {
+            auto eventPacket = packet.asPtr<IEventPacket>(true);
+            if (eventPacket.getEventId() == event_packet_id::DATA_DESCRIPTOR_CHANGED)
+            {
+                auto params = eventPacket.getParameters();
+                DataDescriptorPtr newValueDescriptor = params[event_packet_param::DATA_DESCRIPTOR];
+                DataDescriptorPtr newDomainDescriptor = params[event_packet_param::DOMAIN_DATA_DESCRIPTOR];
 
+                if (newValueDescriptor.assigned())
+                {
+                    dataDescriptor = newValueDescriptor;
+                }
+                if (newDomainDescriptor.assigned())
+                {
+                    domainDescriptor = newDomainDescriptor;
+                }
+            }
+        }
         packet = connection.dequeue();
-    }        
-    
-    info.dataPacket = packet;
+    }
+
+    if (packet.assigned() && packet.getType() == PacketType::Data)
+    {
+        info.dataPacket = packet;
+        info.prevSampleIndex = 0;
+    }
+
+    if (dataDescriptor.assigned() || domainDescriptor.assigned())
```

Comment:
This one will not be able to handle gap packets. I want gap checking enabled on input ports when the reader is connected from signals and properly forwards gap (event) packets.